### DISCUSSION TITLE
Support for UI in non-main thread

### DISCUSSION THREAD

**Discussion by akiross** [slint-ui/slint#1277] [General] [answered]
Hello, there's a little application I'm developing that needs to run the UI off the main thread (the output is a dynamic library that is loaded into another software, a non-blocking entry point is called to perform all the initialization: I have to start the UI there, keep it alive when the init function ends; I'm currently using GTK).

Reading the docs and a few discussions, it seems that slint basically requires using the main thread for creating the UI and its components, but it's possible to run tasks from other threads using `invoke_from_event_loop`. It seems that this choice depends on the backend being used ([wgpu](https://wgpu.rs/) might help here :smile:) - but from the docs it seems there are backends that might not have this restriction, but there are no precise details.

Is the Qt backend able to run inside a non-main thread (on linux)?

A related question: are there plans or intentions to drop this requirement in the future? Personally I'd be very happy to see a portable wgpu backend for slint, but I'm sure that's not in scope for you right now.

Thank you

**Comment by ogoffart** [comment, marked as answer]
The idea is that slint runs in the main thread, and you can then communicate with a thread using message queue, for example.

And example of this is Cargo-UI, if you have not seen it. The idea is that each callback sends message to the thread like so: 
https://github.com/slint-ui/cargo-ui/blob/3e2780dbffc9f123fc5205ffe242805d887f49ad/src/main.rs#L59
Then the thread intercept the message, do the processing, and communicate the result back with `invoke_from_event_loop`, or `Weak::upgrade_in_event_loop` https://github.com/slint-ui/cargo-ui/blob/3e2780dbffc9f123fc5205ffe242805d887f49ad/src/cargo.rs#L513

> Is the Qt backend able to run inside a non-main thread (on linux)?

I do think that works on Linux.  But it does not work on macOs because of platform limitation, and that's also why the winit crate forbid it on all platform to be consistent.
We do plan to support wgpu via femtovg, but that won't help i think, because we still need to integrate with native libraries on some platform (macOS, mobile, ...) just to get a window handle, ant they only allow the GUI eventloop to run in the main thread.
